### PR TITLE
Fixed dependency injection event source tests randomly failing.

### LINKED ISSUE #35753: DependencyInjectionEventSourceTests.EmitsDynamicMethodBuiltEvent test failure on Windows in CI
### Builds
|Build|Pull Request | Test Failure Count|
| --- | --- | --- |
|[#626777](https://dev.azure.com/dnceng/public/_build/results?buildId=626777)|Rolling|1|
|[#626849](https://dev.azure.com/dnceng/public/_build/results?buildId=626849)|#35712|1|
|[#626950](https://dev.azure.com/dnceng/public/_build/results?buildId=626950)|Rolling|1|
### Configurations
- netcoreapp5.0-Windows_NT-Debug-x64-Mono_release-Windows.10.Amd64.Server19H1.ES.Open
- netcoreapp5.0-Windows_NT-Release-x64-Mono_release-Windows.10.Amd64.Server19H1.Open
- netcoreapp5.0-Windows_NT-Release-x86-CoreCLR_release-Windows.10.Amd64.ServerRS5.Open
### Helix Logs
|Build|Pull Request|Console|Core|Test Results|Run Client|
| --- | --- | --- | --- | --- | --- |
|[#626777](https://dev.azure.com/dnceng/public/_build/results?buildId=626777)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/6e37991f-9aa2-4a67-9dd3-cb3bf4243692/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/console.c054715e.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/6e37991f-9aa2-4a67-9dd3-cb3bf4243692/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-50-preview46e37991f9aa24a679d/Microsoft.Extensions.DependencyInjection.Tests/a4bb3d3e-9be2-4ce9-9006-456816cc38f9.log?sv=2019-02-02&se=2020-05-11T18%3A15%3A55Z&sr=c&sp=rl&sig=9V1xYWazICZ6CYS1mvhEfJ8tuNHOt%2FNMgbVs%2BOasPEU%3D)|
|[#626849](https://dev.azure.com/dnceng/public/_build/results?buildId=626849)|#35712|[console.log](https://helix.dot.net/api/2019-06-17/jobs/e03bbe9b-1e74-4de2-9e07-ae9f4867c35e/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/console.83c7b96e.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/e03bbe9b-1e74-4de2-9e07-ae9f4867c35e/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-35712-merge-e03bbe9b1e744de29e/Microsoft.Extensions.DependencyInjection.Tests/1255f0c6-0bf0-4445-a070-0b292cf07e2f.log?sv=2019-02-02&se=2020-05-11T19%3A42%3A07Z&sr=c&sp=rl&sig=%2BAuIuU2VIo0eDTu4unzCju1lWTYUTW83dYwa1pf9wkQ%3D)|
|[#626950](https://dev.azure.com/dnceng/public/_build/results?buildId=626950)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/a23a41ad-f70c-4f9c-b9d3-051b314baa0e/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/console.21d8220f.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/a23a41ad-f70c-4f9c-b9d3-051b314baa0e/workitems/Microsoft.Extensions.DependencyInjection.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-helixqueuesupdate-a23a41adf70c4f9cb9/Microsoft.Extensions.DependencyInjection.Tests/fd04e531-bef8-4205-a39f-e3621e3c615e.log?sv=2019-02-02&se=2020-05-11T21%3A22%3A37Z&sr=c&sp=rl&sig=kvDn%2B6Bvzb9QgTT%2BsV4gwiqCk44fGAhmVjt%2BWFjtHR8%3D)|

```
 Microsoft.Extensions.DependencyInjection.DependencyInjectionEventSourceTests.EmitsDynamicMethodBuiltEvent [FAIL]
      System.InvalidOperationException : Sequence contains more than one matching element
      Stack Trace:
        /_/src/libraries/System.Linq/src/System/Linq/ThrowHelper.cs(22,0): at System.Linq.ThrowHelper.ThrowMoreThanOneMatchException()
        /_/src/libraries/System.Linq/src/System/Linq/Single.cs(74,0): at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
        /_/src/libraries/Microsoft.Extensions.DependencyInjection/tests/DependencyInjectionEventSourceTests.cs(206,0): at Microsoft.Extensions.DependencyInjection.DependencyInjectionEventSourceTests.EmitsDynamicMethodBuiltEvent()
```

Will disable the test in the meantime.

### PR DESCRIPTION
**Pull request by koenigst**
* Filtered events in the listener by the service provider hash. This prevents events from other test classes impacting the event source tests.
* Reused the listener to improve performance.

Fixes #35753

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 1]
```diff
@@ -6,31 +6,17 @@
...
-    [Collection(nameof(EventSourceTests))]
-    public class DependencyInjectionEventSourceTests : IDisposable
-    {
-        private readonly TestEventListener _listener = new TestEventListener();
-
-        public DependencyInjectionEventSourceTests()
-        {
-            // clear the provider list in between tests
-            typeof(DependencyInjectionEventSource).GetField("_providers", BindingFlags.NonPublic | BindingFlags.Instance)
-                .SetValue(DependencyInjectionEventSource.Log, new List<WeakReference<ServiceProvider>>());
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:30]: Do you know what was the causing the random failure? Related to using WeakReference or Dispose?

Were you able to reproduce the test failures?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 2]
```diff
@@ -343,33 +331,77 @@ public void EmitsServiceProviderBuilt()
...
+            }
+        }
 
-            EventWrittenEventArgs serviceProviderBuiltEvent = _listener.EventData.Single(e => e.EventName == "ServiceProviderBuilt");
-            Assert.Equal(1, GetProperty<int>(serviceProviderBuiltEvent, "singletonServices"));
+        private static bool TryGetProperty<T>(EventWrittenEventArgs data, string propName, out T result)
+        {
+            if (data.PayloadNames is { } names &&
+                data.Payload is { } payload &&
+                names.IndexOf(propName) is var index and >= 0 &&
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:359]:
```suggestion
                names.IndexOf(propName) is int index and >= 0 &&
```

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 3]
```diff
@@ -343,33 +331,77 @@ public void EmitsServiceProviderBuilt()
...
+            using var listener = new TestEventListener();
+            listener.EnableEvents(DependencyInjectionEventSource.Log, EventLevel.Verbose);
+            try
+            {
+                EventWrittenEventArgs serviceProviderBuiltEvent = listener.EventDataFor(provider).Single(e => e.EventName == "ServiceProviderBuilt");
+                Assert.Equal(1, GetProperty<int>(serviceProviderBuiltEvent, "singletonServices"));
 
-            _listener.EnableEvents(DependencyInjectionEventSource.Log, EventLevel.Verbose);
+                EventWrittenEventArgs serviceProviderDescriptorsEvent = listener.EventDataFor(provider).Single(e => e.EventName == "ServiceProviderDescriptors");
+                Assert.NotNull(JObject.Parse(GetProperty<string>(serviceProviderDescriptorsEvent, "descriptors")));
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:347]: Is this the only use of Newtonsoft? Can we use STJ instead?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 4]
```diff
@@ -343,33 +331,77 @@ public void EmitsServiceProviderBuilt()
...
+
+            public void Dispose()
+            {
+                Listener.DisableEvents(DependencyInjectionEventSource.Log);
+                Listener.Dispose();
+            }
+        }
 
-        private class TestEventListener : EventListener
+        internal class TestEventListener : EventListener
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:392]: why the change from private to internal here?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 5]
```diff
@@ -343,33 +331,77 @@ public void EmitsServiceProviderBuilt()
...
         }
 
-        private T GetProperty<T>(EventWrittenEventArgs data, string propName)
-            => (T)data.Payload[data.PayloadNames.IndexOf(propName)];
+        private static T GetProperty<T>(EventWrittenEventArgs data, string propName) =>
+            TryGetProperty(data, propName, out T result)
+                ? result
+                : throw new ArgumentException($"No property {propName} with type {typeof(T)} found.");
+
+        public class TestEventListenerFixture : IDisposable
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:375]: Can this be private?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs [hunk 6]
```diff
@@ -239,13 +225,15 @@ public void EmitsScopeDisposedEvent()
...
+            int hashCode = serviceProvider.GetHashCode();
+            DependencyInjectionEventSource.Log.ServiceRealizationFailed(exception, hashCode);
 
             var eventName = nameof(DependencyInjectionEventSource.Log.ServiceRealizationFailed);
-            var serviceRealizationFailedEvent = _listener.EventData.Single(e => e.EventName == eventName);
+            var serviceRealizationFailedEvent = Listener.EventDataFor(serviceProvider).Single(e => e.EventName == eventName);
 
             Assert.Equal("System.Exception: Test error.", GetProperty<string>(serviceRealizationFailedEvent, "exceptionMessage"));
-            Assert.Equal(1234, GetProperty<int>(serviceRealizationFailedEvent, "serviceProviderHashCode"));
+            Assert.Equal(hashCode, GetProperty<int>(serviceRealizationFailedEvent, "serviceProviderHashCode"));
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:236]: What was the reason for the changing the literal `1234` to a an actual hashcode?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs (hunk 1 above)
Comment (koenigst) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:30]: The issue is that the test assumes it is the only one writing to the event source. This is not true because xunit executes the test collections/classes in parallel.

I was able to reproduce the error by inserting `Thread.Sleep` between the test clearing the event source and asserting on the events.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs (hunk 6 above)
Comment (koenigst) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:236]: The service provider needs to have a real hash code for the new code in the `DependencyInjectionEventSource` to work.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs (hunk 4 above)
Comment (koenigst) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:392]: Because the `TestEventListenerFixture` has a property of type `TestEventListener` that needs to visible in the test class.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs (hunk 3 above)
Comment (koenigst) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs:347]: This line used to be on line 361. I did not change it. Git was just unable to track it. I would prefer not having this unrelated change in this PR.

**Comment by steveharter** [comment]
/ba-g Timeout on Build browser-wasm linux Release LibraryTests

**Comment by steveharter** [comment]
/ba-g there is a timeout on Build browser-wasm linux Release LibraryTests

**Review by steveharter — APPROVED** [review]
Thanks!
